src/transport.c

```diff
@@ -524,35 +559,54 @@ int _libssh2_transport_read(LIBSSH2_SESSION * session)
                since it is used for the hash later on. */
             int skip = session->remote.mac->mac_len;
 
+            if(CRYPT_FLAG_R(session, INTEGRATED_MAC))
+                /* This crypto method DOES need the MAC to go through
+                   decryption so it can be authenticated. */
+                skip = 0;
+
             /* if what we have plus numbytes is bigger than the
                total minus the skip margin, we should lower the
                amount to decrypt even more */
-            if((p->data_num + numbytes) > (p->total_num - skip)) {
-                numdecrypt = (p->total_num - skip) - p->data_num;
+            if((p->data_num + numbytes) >= (p->total_num - skip)) {
+                /* decrypt the entire rest of the package */
+                numdecrypt = MAX(0,
+                               (int)(p->total_num - skip) - (int)p->data_num);
+                firstlast = LAST_BLOCK;
             }
             else {
                 ssize_t frac;
                 numdecrypt = numbytes;
                 frac = numdecrypt % blocksize;
                 if(frac) {
-                    /* not an aligned amount of blocks,
-                       align it */
+                    /* not an aligned amount of blocks, align it by reducing
+                       the number of bytes processed this loop */
                     numdecrypt -= frac;
                     /* and make it no unencrypted data
                        after it */
                     numbytes = 0;
                 }
+                if(CRYPT_FLAG_R(session, INTEGRATED_MAC)) {
+                    /* Make sure that we save enough bytes to make the last
+                     * block large enough to hold the entire integrated MAC */
+                    numdecrypt = MIN(numdecrypt, (int)
```

Comment:
```suggestion
                    numdecrypt = LIBSSH2_MIN(numdecrypt, (int)
```
Prefer newly added `LIBSSH2_MAX()`/`LIBSSH2_MIN()`.